contracts/src/Summa.sol

```diff
@@ -5,116 +5,159 @@ pragma solidity ^0.8.18;
+        uint256 amountToProve;
+        bytes[] addresses;
+        bytes balanceRetrieverArgs;
+    }
 
+    event AddressVerifierSet(bytes32 indexed addressType, address verifier);
+    event BalanceRetrieverSet(
+        bytes32 indexed assetType,
+        address balanceRetriever
+    );
+    event ExchangeAddressesSubmitted(OwnedAddress[] addresses);
     event ProofOfSolvencySubmitted(uint256 indexed mstRoot);
 
     constructor(IVerifier _verifier) {
         verifier = _verifier;
     }
 
+    /**
+     * @dev Set the address of the asset balance retriever for a given asset type
+     * @param retriever The address of the asset balance retriever smart contract
+     */
+    function setBalanceRetriever(address retriever) public onlyOwner {
+        require(retriever != address(0), "Invalid balance retriever");
+        IBalanceRetriever _balanceRetriever = IBalanceRetriever(retriever);
+        bytes32 assetType = _balanceRetriever.getAssetType();
+        require(assetType != bytes32(0), "Invalid asset type");
+        balanceRetriever[assetType] = _balanceRetriever;
+        emit BalanceRetrieverSet(assetType, retriever);
+    }
+
+    /**
+     * @dev Set the address of the address ownership verifier for a given asset type
+     * @param _verifier The address of the address ownership verifier smart contract
+     */
+    function setAssetAddressVerifier(address _verifier) public onlyOwner {
+        require(_verifier != address(0), "Invalid address verifier");
+        IAddressOwnershipVerifier addressOwnershipVerifier = IAddressOwnershipVerifier(
+                _verifier
+            );
+        bytes32 addressType = addressOwnershipVerifier.getAddressType();
+        require(addressType != bytes32(0), "Invalid address type");
+        cexAddressVerifier[addressType] = addressOwnershipVerifier;
+        emit AddressVerifierSet(addressType, _verifier);
+    }
+
     function submitProofOfAccountOwnership(
-        address[] memory _cexAddresses,
-        bytes[] memory cexSignatures,
-        string memory message
+        OwnedAddress[] memory _cexAddresses
     ) public {
-        require(
-            _cexAddresses.length == cexSignatures.length &&
-                _cexAddresses.length > 0,
-            "CEX addresses and signatures count mismatch"
-        );
-
         for (uint i = 0; i < _cexAddresses.length; i++) {
-            if (i >= cexAddresses.length) {
-                cexAddresses.push(_cexAddresses[i]);
-            } else if (_cexAddresses[i] != cexAddresses[i]) {
-                cexAddresses[i] = _cexAddresses[i];
-            }
-            address recoveredPubKey = keccak256(abi.encode(message))
-                .toEthSignedMessageHash()
-                .recover(cexSignatures[i]);
+            bytes32 addressHash = keccak256(_cexAddresses[i].cexAddress);
+            require(!confirmedAddress[addressHash], "Address already verified");
+            confirmedAddress[addressHash] = true;
```

Comment:
No, the transaction will revert entirely. Also, it's a good practice to modify state before the external call to avoid any hacks based on malicious external contracts, e.g., reentrancy (although we probably shouldn't worry because `verifyAddressOwnership` shouldn't be ever controlled by a malicious actor).